### PR TITLE
Add bunzz upload command for uploading to SCH

### PR DESCRIPTION
**Pull request by shahincsejnu**

### CODE REVIEW

File: src/commands/upload.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+interface HardhatConfig {
+  paths: {
+    sources: string;
+    artifacts: string;
+  };
+  solidity: any;
+}
+
+async function findSolidityFiles(dir: string): Promise<string[]> {
```
Comment (RaresGeo) [inline src/commands/upload.ts:28]: Nice function, good job!

File: src/commands/upload.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    } else if (fullPath.endsWith(".sol")) {
+      files.push(fullPath);
+    }
+  }
+
+  return files;
+}
+
+function parseImports(fileContent: string): string[] {
+  const importRegex = /^\s*import\s+(["'])(.*?)\1;/gm;
```
Comment (RaresGeo) [inline src/commands/upload.ts]: There are two kinds of imports in Solidity. One that just imports an entire file (i.e. `
import "../utils/ContextUpgradeable.sol";`), or ones that import specific things from a file. (i.e. `import {Context} from "../utils/Context.sol";`)

Consider upgrading to this regex rule: `/import\s+(?:([\w\s{},*]+)\s+from\s+)?(['"])(.*?)(['"]);/g` and updating the rest of the code accordingly.

File: src/commands/upload.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  processedFiles.add(filePath);
+
+  const content = await fetchFileContent(filePath);
+  const imports = parseImports(content);
+
+  let files: SolidityFile[] = [{ path: filePath, content }];
+
+  for (const importPath of imports) {
+    // Skip third-party imported files
+    if (importPath.startsWith("@") || !importPath.startsWith(".")) {
```
Comment (RaresGeo) [inline src/commands/upload.ts]: I'm not sure about what files need to be excluded specifically but can you please test that this works with [this contract](https://app.bunzz.dev/decipher/chains/1/addresses/0xb258ca5559b11cd702f363796522b04d7722ea56)? It uses an imported library solmate. Where solmate is imported, it looks like this: `import "solmate/utils/SignedWadMath.sol";` so no `'@'` and no `'.'` at the start of the string. I'm not sure if you want to take this with you or ignore it, but please make sure it works.

File: src/commands/upload.ts [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+      allFiles = allFiles.concat(processed);
+      processed.forEach((f) => processedFiles.add(f.path)); // Add to processed set
+    }
+  }
+
+  return allFiles;
+}
+
+async function getContractNames(content: string): Promise<string[]> {
+  const contractNameRegex = /^\s*contract\s+(\w+)\s+/gm;
```
Comment (RaresGeo) [inline src/commands/upload.ts:121]: 👍

File: src/commands/upload.ts [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  const questions = [
+    {
+      type: "list",
+      name: "selectedContract",
+      message: "Please select a contract as the base contract:",
+      choices: contractNames,
+    },
+  ];
+
+  const inquirer = (await import("inquirer")).default;
```
Comment (RaresGeo) [inline src/commands/upload.ts]: Why import it dynamically instead of statically?

File: src/commands/upload.ts [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+      "The uploading process requires hardhat configuration file. Please run `bunzz init` to create it. Detail: https://docs..."
+    );
+  }
+
+  const configFileContents = fs.readFileSync(hardhatConfigPath, "utf8");
+  let hardhatConfig: HardhatConfig;
+
+  try {
+    // Evaluate the config file in a VM to safely obtain the exports
+    const script = new vm.Script(configFileContents, {
```
Comment (RaresGeo) [inline src/commands/upload.ts:189]: Interesting approach. Great work! I'm curious how you found out this is a solution

File: src/commands/upload.ts [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    // get the artifacts of base contract
+    const { ABI, bytecode } = getArtifacts(projectPath, rootContractName);
+
+    // validate the artifacts and respective code
+
+    // store the artifacts and related infos to the bunzz server
+    console.log("abi:", ABI);
+    console.log("bytecode:", bytecode);
```
Comment (RaresGeo) [inline src/commands/upload.ts:291]: Maybe we should remove these in the future, you can keep them for now since it's not a PR into `main`

**Review by RaresGeo — APPROVED** [review]
Approved, but please answer the questions I left.

File: src/commands/upload.ts (hunk 7 above)
Comment (shahincsejnu) [reply src/commands/upload.ts:291]: yeah, later we'll remove those and use for sending to bunzz server purpose

File: src/commands/upload.ts (hunk 3 above)
Comment (shahincsejnu) [reply src/commands/upload.ts]: @RaresGeo thank you for the example, basically as bunzz SCH you can see we don't put the @oppenzeplin codes, so I'll then only gonna ignore which start with `@` and which are not in project path! later if we need to ignore/include more can be update!

File: src/commands/upload.ts (hunk 2 above)
Comment (shahincsejnu) [reply src/commands/upload.ts]: okay updated the regex to this: `/import(?:.*?from\s+)?(['"])([^'"]+)(['"]);/g`, it captures the second type of import!

**Pull request merged**